src/tests/ftest/util/dmg_utils.py

```diff
@@ -1,52 +1,103 @@
+    """Defines a object representing a dmg (or daos_shell) command."""
+
+    def __init__(self, path):
+        """Create a dmg Command object."""
+        super(DmgCommand, self).__init__(path, "daos_shell")
+
+        self.request = BasicParameter("{}")
+        self.action = BasicParameter("{}")
+
+        # daos_shell options
+        self.hostlist = FormattedParameter("-l {}")
+        self.hostfile = FormattedParameter("-f {}")
+        self.configpath = FormattedParameter("-o {}")
+        self.insecure = FormattedParameter("-i", None)
+
+        # dmg options
+        self.gid = FormattedParameter("--gid={}")
+        self.uid = FormattedParameter("--uid={}")
+        self.group = FormattedParameter("--group={}")
+        self.mode = FormattedParameter("--mode={}")
+        self.size = FormattedParameter("--size={}")
+        self.nvme = FormattedParameter("--nvme={}")
+        self.svcn = FormattedParameter("--svcn={}")
+        self.target = FormattedParameter("--target={}")
+        self.force = FormattedParameter("--force", False)
+        self.pool = FormattedParameter("--pool={}")
+        self.svc = FormattedParameter("--svc={}")
+        self.rank = FormattedParameter("--rank={}")
+        self.cont = FormattedParameter("--cont={}")
+        self.oid = FormattedParameter("--oid={}")
+
+    def get_param_names(self):
+        """Get a sorted list of daos_server command parameter names."""
+        names = self.get_attribute_names(FormattedParameter)
+        names.extend(["request", "action"])
+        return names
+
+    def get_params(self, test, path="/run/dmg/*"):
+        """Get values for all of the dmg command params using a yaml file.
+
+        Sets each BasicParameter object's value to the yaml key that matches
+        the assigned name of the BasicParameter object in this class. For
+        example, the self.block_size.value will be set to the value in the yaml
+        file with the key 'block_size'.
+
+        Args:
+            test (Test): avocado Test object
+            path (str, optional): yaml namespace. Defaults to "/run/dmg/*".
+
+        """
+        super(DmgCommand, self).get_params(test, path)
+
+    def run(self, timeout=None, verbose=True, env=None, sudo=False):
+        """ Run the dmg command.
+
+        Args:
+            timeout (int, optional): timeout in seconds. Defaults to None.
+            verbose (bool, optional): display command output. Defaults to True.
+            env (dict, optional): env for the command. Defaults to None.
+            sudo (bool, optional): sudo will be prepended to the command.
+                Defaults to False.
+
+        Raises:
+            process.CmdError: Avocado command exception
+
+        Returns:
+            process.CmdResult: CmdResult object containing the results from
+            the command.
+
+        """
+        return process.run(self.__str__(), timeout, verbose, env=env,
+                           shell=True, sudo=sudo)
```

Comment:
Seems like this would be a standard `CommandWithParameters()` method.  I don't see anything specific to `DmgCommand()` here.

Reply:
Originally the run method was part of CommandWithParameters, but after some discussion with Phil we decided to include it just in dmg. I think its worth rethinking to put it back in CommandWithParameters.

Reply:
The current implementations of `CommandWithParameters` do all have a `run(self, manager, attach_info, processes, hostfile, display=True)` method (we could rename `SeverCommand`'s `start()` to `run()`), but in order for there to be a base `run(self)` method the `MdtestCommand` and `IorCommand` classes would need a different mechanism for setting the `manager, attach_info, processes, hostfile, display=True` attributes.  Much of this could be handled by making them attributes to their respective class, but I think a cleaner solution would be to create `CommandWithParameter` classes for `mpirun` and `otrerun` who's attributes would encompass the current inputs for the `IorCommand.run()` method.  This way only one additional attribute would be needed for the `MdtestCommand` and `IorCommand` classes using a `run(self)` method.

Reply:
@phender Sure.  That sounds good.  But seems like that needs a different PR though.

But even in the meanwhile, it doesn't seem unreasonable to push down the generic `run()` here from `DmgCommand()` into `CommandWithParameters()`, knowing that children of `CommandWithParameters()` can completely override the `run()` method if they need something more complicated.